jobs/jobs/utils.py

```diff
@@ -585,7 +590,147 @@ async def get_job_progress(
+    if job.mode == JobMode.Manual:
+        await update_manual_job_status(session, job, response)
+        return response
+
+    elif job.mode == JobMode.Automatic:
+        return await handle_automatic_job(
+            session, job, response, job_id, api_client
+        )
+
+    logger.warning(
+        f"Unknown job mode: {job.mode}, treating as manual by default."
+    )
+    await update_manual_job_status(session, job, response)
+    return response
+
+
+async def update_manual_job_status(
+    session: Session, job: CombinedJob, response: Dict[str, int]
+) -> None:
+    """Updates the job status based on the DAG execution response."""
+    if response.get("finished", 0) > 0:
+        if response["finished"] == response["total"]:
+            db_service.update_job_status(session, job, Status.finished)
+        else:
+            db_service.update_job_status(session, job, Status.in_progress)
+
+
+async def handle_automatic_job(
+    session: Session,
+    job: CombinedJob,
+    response: Dict[str, int],
+    job_id: int,
+    api_client: client.ApiClient,
+) -> Dict[str, int]:
+    """Handles automatic jobs with or without a pipeline."""
+    if response.get("total", 0) == 0:
+        return await handle_pipeline_driven_job(
+            session, job, response, job_id, api_client
+        )
+    else:
+        await update_manual_job_status(session, job, response)
+        return response
+
+
+async def handle_pipeline_driven_job(
+    session: Session,
+    job: CombinedJob,
+    response: Dict[str, int],
+    job_id: int,
+    api_client: client.ApiClient,
+) -> Dict[str, int]:
+    """Handles automatic jobs that rely on pipeline DAG execution."""
+    pipeline_id = job.pipeline_id
+    dag_run_id = f"{pipeline_id}:{job_id}"
+
+    await activate_pipline(pipeline_id, api_client)
+    db_service.update_job_status(session, job, Status.in_progress)
+
+    await wait_for_pipline_completion(
+        session, job, pipeline_id, dag_run_id, api_client
+    )
+
+    pipeline_result = await get_pipline_status(
+        session, job, job_id, pipeline_id, api_client
+    )
+
+    progress_bar_value = (
+        1 if pipeline_result == AirflowPiplineStatus.success.value else 0
+    )
+    response["total"] = 1
+    response["finished"] = progress_bar_value
+
+    new_status = Status.finished if progress_bar_value == 1 else Status.failed
+    db_service.update_job_status(session, job, new_status)
+
+    return response
+
+
+async def get_pipline_status(
+    session: Session,
+    job: CombinedJob,
+    job_id: int,
+    pipeline_id: str,
+    api_client: client.ApiClient,
+) -> str:
+    try:
+        pipeline_result = await airflow_utils.get_dag_status(
+            pipeline_id, job_id, api_client
+        )
+    except RuntimeError as e:
+        logger.exception(
+            "Runtime error occurred while getting DAG status result"
+        )
+        db_service.update_job_status(session, job, Status.failed)
+        raise fastapi.HTTPException(
```

Comment:
Do not expose internals to the user. Consider HTTPExceptions as messages to the end user. In your case something like "Failed to fetch pipeline status" is enough. User doesn't even know what dag is, for them it is a job with pipelines sometimes